Nodejs/Product/TestAdapter/TestFilesUpdateWatcher.cs

```diff
@@ -43,84 +36,47 @@ public bool AddDirectoryWatch(string path)
         {
             ValidateArg.NotNull(path, "path");
 
-            if (!string.IsNullOrEmpty(path))
+            if (!this.watchedFiles.ContainsKey(path) && ErrorHandler.Succeeded(this.fileWatcher.AdviseDirChange(path, VSConstants.S_OK, this, out uint cookie)))
             {
-                if (!_fileWatchers.ContainsKey(path) && Directory.Exists(path))
-                {
-                    var watcher = new FileSystemWatcher(path);
-                    _fileWatchers[path] = watcher;
-
-                    watcher.IncludeSubdirectories = true;
-                    watcher.Changed += OnChanged;
-                    watcher.Renamed += OnRenamed;
-                    watcher.EnableRaisingEvents = true;
-                    return true;
-                }
+                this.watchedFiles.Add(path, cookie);
+                return true;
             }
             return false;
         }
 
-        public void RemoveWatch(string path)
+        public bool RemoveWatch(string path)
         {
             ValidateArg.NotNull(path, "path");
 
-            if (!string.IsNullOrEmpty(path))
+            if (this.watchedFiles.TryGetValue(path, out uint cookie))
             {
-                FileSystemWatcher watcher;
-
-                if (_fileWatchers.TryGetValue(path, out watcher))
+                this.watchedFiles.Remove(path);
+                if (Path.HasExtension(path))
                 {
-                    watcher.EnableRaisingEvents = false;
-
-                    _fileWatchers.Remove(path);
-
-                    watcher.Changed -= OnChanged;
-                    watcher.Dispose();
+                    return ErrorHandler.Succeeded(this.fileWatcher.UnadviseFileChange(cookie));
+                }
+                else
+                {
+                    return ErrorHandler.Succeeded(this.fileWatcher.UnadviseDirChange(cookie));
                 }
             }
+            return false;
         }
 
-        private void OnChanged(object sender, FileSystemEventArgs e)
-        {
-            var evt = FileChangedEvent;
-            if (evt != null)
-            {
-                evt(sender, new TestFileChangedEventArgs(null, e.FullPath, TestFileChangedReason.Changed));
-            }
-        }
-
-        private void OnRenamed(object sender, RenamedEventArgs e)
+        int IVsFileChangeEvents.FilesChanged(uint cChanges, string[] rgpszFile, uint[] rggrfChange)
         {
-            var evt = FileChangedEvent;
-            if (evt != null)
+            for (var i = 0; i < cChanges; i++)
             {
-                evt(sender, new TestFileChangedEventArgs(null, e.FullPath, TestFileChangedReason.Renamed));
+                FileChangedEvent?.Invoke(this, new TestFileChangedEventArgs(null, rgpszFile[i], TestFileChangedReason.Changed));
```

Comment:
The old code had distinct events for Changed and Renamed. The new doesn't. Does this change in behavior cause any issues to subscribers?

Reply:
there was a subtle difference in behavior.